src/Database/PostgreSQL/Consumers/Components.hs

```diff
@@ -325,14 +344,58 @@ spawnDispatcher ConsumerConfig{..} cs cid semaphore
           ]
         return (ccJobIndex job, Failed action)
 
+    updateJob :: (idx, Result) -> m ()
+    updateJob (idx, result) = runDBT cs ts $ do
+      now <- currentTime
+      runSQL_ $ case result of
+          Ok Remove -> deleteQuery
+          Failed Remove -> deleteQuery
+          _ -> retryQuery now (isSuccess result) (getAction result)
+      where
+        deleteQuery = "DELETE FROM" <+> raw ccJobsTable <+> "WHERE" <+> raw idxRow <+> "<=" <?> idx
+
+        retryQuery now success action = smconcat
+          [ "UPDATE" <+> raw ccJobsTable <+> "SET"
+          , "  reserved_by = NULL"
+          , ", run_at = CASE"
+          , "    WHEN FALSE THEN run_at"
+          ,      retryToSQL
+          , "    ELSE NULL" -- processed
+          , "  END"
+          , if success then smconcat
+              [ ", finished_at = CASE"
+              , "    WHEN id =" <?> idx <+> "THEN" <?> now
+              , "    ELSE NULL"
+              , "  END"
+              ]
+              else ""
+          , "WHERE" <+> raw idxRow <+> "<=" <?> idx
+          ]
+          where
+            retryToSQL = case action of
+              RerunAfter int -> "WHEN id =" <?> idx <+> "THEN" <?> now <+> "+" <?> int
+              RerunAt time -> "WHEN id =" <?> idx <+> "THEN" <?> time
+              MarkProcessed -> ""
+              Remove -> error "updateJob: Remove should've been filtered out"
+
+        idxRow = case ccMode of
+          Standard -> error $ "'updateJob' should never be called when ccMode = " <> show Standard
```

Comment:
If you just say "Standard" in the error message, you can save `<>`, `show` and `$` 😉

Reply:
I did this on purpose so that if anyone changes the data constructor in the future this won't compile